### PR TITLE
Protein terminal mods implementation in Tide

### PR DESCRIPTION
**Pull request by rita-gwen**

### CODE REVIEW

**Comment by rita-gwen** [comment]
Re. issue 469, I think it is not possible to make tide-search output indicators for the terminal mods either in sequence or in modspec columns in the same way comet does. There is no information in the index that would allow to reference the initial modspec. All it keeps is the list of mod masses.
A acceptable fix would be to make sure that PepXmlWriter can parse such modspecs correctly.

File: src/app/LocalizeModification.cpp [hunk 1]
```diff
@@ -129,6 +129,7 @@ int LocalizeModificationApplication::main(int argc, char** argv) {
     VariableModTable* modTable = getModTable(match);
     MassConstants::Init(modTable->ParsedModTable(),
                         modTable->ParsedNtpepModTable(), modTable->ParsedCtpepModTable(),
+                        nullptr, nullptr,
```
Comment (rita-gwen) [inline src/app/LocalizeModification.cpp:132]: Comment by bmx8177: In src/app/LocalizeModification.cpp at line 130 you have two null pointers as input into MassConstants::init. Is this b/c you do not need to localize Nterm and Cterm mods?

Comment (rita-gwen) [reply src/app/LocalizeModification.cpp:132]: Correct. It also seems that LocalizeModifications is not used much, so figuring it out is a lower priority.

File: src/app/ReadTideIndex.cpp [hunk 2]
```diff
@@ -52,7 +52,10 @@ int ReadTideIndex::main(int argc, char** argv) {
   }
 
   const pb::Header::PeptidesHeader& pepHeader = peptides_header.peptides_header();
-  MassConstants::Init(&pepHeader.mods(), &pepHeader.nterm_mods(), &pepHeader.cterm_mods(),
+  //TODO: replace nullptr with actual tables.
```
Comment (rita-gwen) [inline src/app/ReadTideIndex.cpp:55]: Comment by bmx8177: In src/app/ReadTideIndex.cpp at line you have a comment at line 55 about removing nullptrs. I don't see those nullptrs. Is this a stray comment?

Comment (rita-gwen) [reply src/app/ReadTideIndex.cpp:55]: Yes, I replaced the null pointers, but did not remove the comment. It should be removed.

File: src/app/SubtractIndexApplication.cpp [hunk 3]
```diff
@@ -69,9 +69,14 @@ int SubtractIndexApplication::main(int argc, char** argv) {
       TideSearchApplication::PROTEIN_LEVEL_DECOYS = true;
     }
   }
+  //TODO: add protein terminal mods to the header protobuf definition
+  //TODO: replace nullptr with actual tables.
```
Comment (rita-gwen) [inline src/app/SubtractIndexApplication.cpp:73]: Same here.

Comment (bmx8177) [reply src/app/SubtractIndexApplication.cpp:73]: Should line 72 be removed as well? I

**Comment by bmx8177** [comment]
Any idea why line 562, 615, and 600 differ in src/app/TideIndexApplication.cpp? Two call calcMassPepTide and the other line does not.

**Comment by bmx8177** [comment]
In src/app/tide/peptide_mods3.cc there seems to be some inconsistency between the function OutputCtermMods and OutputNtermMods. In the Nterm version there are three for loops while in the CTerm version there are two loops. The difference appears to be that you combined two of the for loop in the CTerm version that you did not combine in the NTerm version. Is that what you wanted to do?

**Comment by bmx8177** [comment]
I do not see the nterm-protein-mods-spec and the cterm-protein-mods-spec option in the tide-index webpage. You will need to edit the Param.cpp file for that.

**Comment by bmx8177** [comment]
/net/gs/vol1/home/linandy/projects/winterRotation/crux-toolkit/doc/user/release-notes.html needs to be updated with this.

**Comment by rita-gwen** [comment]
There is no change in src/app/TideIndexApplication.cpp on line 600. calcPepMassTide signature has changed, so it's invocation changed as well on lines 506, 568, 615 in the new version. It would be easier if you attach your comments directly to the relevant lines of code.

**Comment by rita-gwen** [comment]
Yes, I merged the first two loops in OutputCtermMods because they were identical and differed only in the AA residue character. I will do it in OutputNtermMods for consistency as well.

**Comment by rita-gwen** [comment]
I'm still working on the documentation changes, they will be in the next commit.

**Pull request merged**